async_fuse/src/byte_slice.rs

```diff
@@ -0,0 +1,208 @@
+        ByteSlice { data }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub(crate) fn fetch_all(&mut self) -> &'a [u8] {
+        mem::replace(&mut self.data, &[])
+    }
+
+    pub(crate) fn fetch_bytes(&mut self, amt: usize) -> anyhow::Result<&'a [u8]> {
+        if amt > self.data.len() {
+            anyhow::bail!(
+                "no enough bytes to fetch, remaining {} bytes but to fetch {} bytes",
+                self.data.len(),
+                amt
+            );
+        }
+        let (bytes, remain) = self.data.split_at(amt);
+        self.data = remain;
+        Ok(bytes)
+    }
+
+    pub(crate) fn fetch<T: FuseAbiData>(&mut self) -> anyhow::Result<&'a T> {
+        let elem_len: usize = mem::size_of::<T>();
+
+        if (self.data.as_ptr() as usize) % mem::align_of::<T>() != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to type {}, pointer={:p} is not a multiple of alignment={}",
+                std::any::type_name::<T>(),
+                self.data.as_ptr(),
+                mem::align_of::<T>()
+            );
+        }
+
+        let bytes: &[u8] = self.fetch_bytes(elem_len).with_context(|| {
+            format!(
+                "failed to build FUSE request payload type {}",
+                std::any::type_name::<T>(),
+            )
+        })?;
+
+        let ret: &T = unsafe { transmute_ref_unchecked(bytes) };
+        Ok(ret)
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn fetch_all_as_slice<T: FuseAbiData>(&mut self) -> anyhow::Result<&'a [T]> {
+        let elem_len: usize = mem::size_of::<T>();
+
+        if self.data.len() % elem_len != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to a slice of type={}, \
+                 the total bytes length={} % the type size={} is nonzero",
+                std::any::type_name::<T>(),
+                self.data.len(),
+                elem_len
+            );
+        }
+
+        if (self.data.as_ptr() as usize) % mem::align_of::<T>() != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to a slice of type={}, \
+                 pointer={:p} is not a multiple of alignment={}",
+                std::any::type_name::<T>(),
+                self.data.as_ptr(),
+                mem::align_of::<T>()
+            );
+        }
+
+        let bytes: &[u8] = self.fetch_all();
+        let ret = unsafe { transmute_slice_unchecked(bytes) };
+        Ok(ret)
+    }
+
+    pub(crate) fn fetch_c_str(&mut self) -> anyhow::Result<&'a CStr> {
+        let strlen: usize = match memchr::memchr(b'\0', self.data) {
+            Some(nul_pos) => nul_pos + 1,
+            None => anyhow::bail!("no trailing zero in bytes, cannot fetch c-string"),
+        };
+        let bytes: &[u8] = self.fetch_bytes(strlen)?;
+        let ret: &CStr = unsafe { CStr::from_bytes_with_nul_unchecked(bytes) };
+        Ok(ret)
+    }
+
+    #[cfg(unix)]
```

Comment:
How about `#[cfg(any((target_os="linux"), (target_os="macos")))]`?
so as to explicitly list supported OS.

Reply:
done